### ISSUE TITLE
[Bug] Metric Tensor function only works with `interface="autograd"`

### ISSUE THREAD

**Issue by jackaraz** [PennyLaneAI/pennylane#2040]
Hi, I'm trying to adapt `qml.QNGOptimizer` to TensorFlow backend however I realised that `qml.metric_tensor` only works for `autograd` backend. As a demonstration below I used a simple circuit
```python
dev = qml.device("default.qubit", wires = 2, shots=None)
@qml.qnode(dev, diff_method="parameter-shift", interface="tf")
def circuit(inputs, weights):
    qml.AngleEmbedding(inputs, wires = range(2), rotation="Y")

    qml.RY(weights[0], wires=0)
    qml.RY(weights[1], wires=1)
    qml.CNOT(wires = [0, 1])

    return qml.probs(op=qml.PauliZ(1))

metric_tensor_fn = qml.metric_tensor(circuit, approx="block-diag")
```
In order to work with TensorFlow-Keras backend `metric_tensor_fn` needs to be able to take `tf.Tensor` and/or `tf.Variable`. However, I realised that when `tf.Tensor` type of object is given to `metric_tensor_fn`
```python
inpt = tf.random.uniform((2,))
weight = tf.Variable(tf.random.uniform((2,)), trainable=True)
metric_tensor_fn(inpt, weight)
```
`cov_matrix` in `/pennylane/math/quantum.py` receives an empty list of observables hence `cov = diag(variances)` raises an error due to an empty list of variances. When the input is `np.ndarray`, on the other hand, this portion works but it fails later on when `pennylane/transforms/classical_jacobian.py` attempts to take the jacobian via TensorFlow backend since the input is not `tf.Tensor` it fails.

PS: this is probably happening because `circuit.tape.graph.trainable_params` does not interpret `tf.Variable` as a trainable object.
```python
circuit(inpt, weight)
circuit.tape.graph.trainable_params
# Output: []
```

Is there a possible way to adapt this function to TensorFlow?

**System settings**
- PennyLane v0.20.0
- TensorFlow v2.5.1

Thanks

**Comment by josh146** [comment]
Hi @jackaraz, thanks for reporting this! This is a bug we discovered the other day, so good to have it in the issue tracker.

The issue arises because PennyLane uses TensorFlow's `tensorflow.python.eager.tape.should_record_backprop` function to determine trainable parameters. In TensorFlow versions <=2.3, this would properly track all `Variable` objects; however was changed in TensorFlow >=2.4 to only track `Variable` objects within a tape context.

Luckily, this means that there is a workaround at the moment: simply place the computation within a `GradientTape`, and TensorFlow will properly communicate the trainable parameters to PennyLane:

```python
inpt = tf.random.uniform((2,), dtype=tf.float64)
weight = tf.Variable(tf.random.uniform((2,), dtype=tf.float64))

with tf.GradientTape() as tape:
    mt = metric_tensor_fn(inpt, weight)

>>> print(mt)
(<tf.Tensor: shape=(2, 2), dtype=float64, numpy=
array([[0.25, 0.  ],
       [0.  , 0.25]])>,)
```

Longer term, how to best fix this on the PennyLane side is still to be determined. The best approach at the moment looks like simply adding a `argnum` argument to `metric_tensor`, so that the trainable arguments can be provided explicitly, e.g.,

```python
>>> qml.metric_tensor(circuit, argnum=1)
```

**Comment by jackaraz** [comment]
Thanks @josh146 this indeed works (also tested with the latest TensorFlow 2.7.0, just so you know).

**Comment by CatalinaAlbornoz** [comment]
I'm glad to hear that this works for you @jackaraz ! Thank you for reporting the bug and the successful test.
